File: ASSIGNMENT10/python/Q.5.py
class TreeNode:
    def __init__(self, key):
        self.left = None
        self.right = None
        self.value = key

class BinarySearchTree:
    def __init__(self):
        self.root = None

    def insert(self, key):
        if self.root is None:
            self.root = TreeNode(key)
        else:
            self._insert_recursive(self.root, key)

    def _insert_recursive(self, node, key):
        if key < node.value:
            if node.left is None:
                node.left = TreeNode(key)
            else:
                self._insert_recursive(node.left, key)
        else:
            if node.right is None:
                node.right = TreeNode(key)
            else:
                self._insert_recursive(node.right, key)

    def inorder_traversal(self, node, result):
        if node is not None:
            self.inorder_traversal(node.left, result)
            result.append(node.value)
            self.inorder_traversal(node.right, result)

def sort_numbers_with_bst(numbers):
    bst = BinarySearchTree()
    
    # Insert all numbers into the BST
    for number in numbers:
        bst.insert(number)

    # Retrieve the sorted numbers
    sorted_numbers = []
    bst.inorder_traversal(bst.root, sorted_numbers)
    return sorted_numbers

# Example Usage
if __name__ == "__main__":
    numbers = [35, 10, 15, 30, 5, 50, 40]
    
    sorted_numbers = sort_numbers_with_bst(numbers)

    # Print results
    print("Original Numbers:", numbers)
    print("Sorted Numbers:", sorted_numbers)

File: ASSIGNMENT10/c/Q.5.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node *left;
    struct Node *right;
};

struct Node* insert(struct Node* node, int data) {
    if (node == NULL) {
        return createNode(data);
    }
    if (data < node->data) {
        node->left = insert(node->left, data);
    } else {
        node->right = insert(node->right, data);
    }
    return node;
}

void inorder(struct Node* node) {
    if (node == NULL) return;
    inorder(node->left);
    printf("%d ", node->data);
    inorder(node->right);
}

int main() {
    int numbers[] = {40, 10, 30, 20, 50, 60};
    int n = sizeof(numbers) / sizeof(numbers[0]);
    
    struct Node* root = NULL;
    for (int i = 0; i < n; i++) {
        root = insert(root, numbers[i]);
    }

    printf("Sorted numbers: ");
    inorder(root);
    printf("\n");
    return 0;
}
